presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java

```diff
@@ -61,66 +51,51 @@
     {
         KafkaTableHandle kafkaTableHandle = (KafkaTableHandle) table;
-        try {
-            SimpleConsumer simpleConsumer = consumerManager.getConsumer(selectRandom(nodes));
-
-            TopicMetadataRequest topicMetadataRequest = new TopicMetadataRequest(ImmutableList.of(kafkaTableHandle.getTopicName()));
-            TopicMetadataResponse topicMetadataResponse = simpleConsumer.send(topicMetadataRequest);
-
-            ImmutableList.Builder<ConnectorSplit> splits = ImmutableList.builder();
-
-            for (TopicMetadata metadata : topicMetadataResponse.topicsMetadata()) {
-                for (PartitionMetadata part : metadata.partitionsMetadata()) {
-                    log.debug("Adding Partition %s/%s", metadata.topic(), part.partitionId());
-
-                    BrokerEndPoint leader = part.leader();
-                    if (leader == null) {
-                        throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Leader election in progress for Kafka topic '%s' partition %s", metadata.topic(), part.partitionId()));
-                    }
-
-                    HostAddress partitionLeader = HostAddress.fromParts(leader.host(), leader.port());
-
-                    SimpleConsumer leaderConsumer = consumerManager.getConsumer(partitionLeader);
-                    // Kafka contains a reverse list of "end - start" pairs for the splits
-
-                    long[] offsets = findAllOffsets(leaderConsumer, metadata.topic(), part.partitionId());
-
-                    for (int i = offsets.length - 1; i > 0; i--) {
-                        KafkaSplit split = new KafkaSplit(
-                                metadata.topic(),
+        try (KafkaConsumer<byte[], byte[]> kafkaConsumer = kafkaConsumerFactory.create()) {
+            List<PartitionInfo> partitionInfos = kafkaConsumer.partitionsFor(kafkaTableHandle.getTopicName());
+
+            List<TopicPartition> topicPartitions = partitionInfos.stream()
+                    .map(KafkaSplitManager::toTopicPartition)
+                    .collect(toImmutableList());
+
+            Map<TopicPartition, Long> partitionBeginOffsets = kafkaConsumer.beginningOffsets(topicPartitions);
+            Map<TopicPartition, Long> partitionEndOffsets = kafkaConsumer.endOffsets(topicPartitions);
+
+            ImmutableList.Builder<KafkaSplit> splits = ImmutableList.builder();
+            Optional<String> keyDataSchemaContents = kafkaTableHandle.getKeyDataSchemaLocation()
+                    .map(KafkaSplitManager::readSchema);
+            Optional<String> messageDataSchemaContents = kafkaTableHandle.getMessageDataSchemaLocation()
+                    .map(KafkaSplitManager::readSchema);
+            for (PartitionInfo partitionInfo : partitionInfos) {
+                TopicPartition topicPartition = toTopicPartition(partitionInfo);
+                HostAddress leader = HostAddress.fromParts(partitionInfo.leader().host(), partitionInfo.leader().port());
```

Comment:
are there more nodes containing topic? For large topics we could distribute load such way